Add Auto-close on inbound/outbound update_fee delay

ariard opened the pull request:
Auto-close timer, if the channel is outbound, we sent a `update_fee`, and we didn't
receive a RAA from counterparty committing this state after `AUTOCLOSE_TIMEOUT` periods,
this channel must be force-closed.
If the channel is inbound, it has been observed the channel feerate should increase,
and we didn't receive a RAA from counterparty committing an `update_fee` after
`AUTOCLOSE_TIMEOUT` periods, this channel must be force-closed.

Close #993 

- [ ] test coverage
- [ ] distrust inbound `update_fee`, verify it's going up
- [x]  if feerates are going downwards, do not trigger timers
- [ ] verify other implems timers to fine-tune `AUTOCLOSE_TIMEOUT` value

TheBlueMatt commented:
Meeting notes - @ariard is going to make this a bit more conservative (eg require the feerate needs to increase 25-50% *and* be higher than our background estimate before force-closing) as well as check what other implementations are doing.

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -626,6 +626,7 @@ const CHECK_CLTV_EXPIRY_SANITY: u32 = MIN_CLTV_EXPIRY_DELTA as u32 - LATENCY_GRA
 #[allow(dead_code)]
 const CHECK_CLTV_EXPIRY_SANITY_2: u32 = MIN_CLTV_EXPIRY_DELTA as u32 - LATENCY_GRACE_PERIOD_BLOCKS - 2*CLTV_CLAIM_BUFFER;
 
+
```
nit

valentinewallace commented on lightning/src/ln/channel.rs:
```diff
@@ -392,6 +398,10 @@ pub(super) struct Channel<Signer: Sign> {
 
+	// Auto-close timer, if the channel is outbound, and we didn't receive a RAA for counterparty
+	// commitment transaction after `AUTOCLOSE_TIMEOUT` periods, this channel must be force-closed.
+	autoclose_timer: u16,
```
nit: seems like it could be a u8

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2660,6 +2667,23 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 		(retain_channel, NotifyOption::DoPersist, ret_err)
 	}
 
+	fn check_channel_autoclose(&self, short_to_id: &mut HashMap<u64, [u8; 32]>, chan_id: &[u8; 32], chan: &mut Channel<Signer>) -> (bool, NotifyOption, Result<(), MsgHandleErrInternal>) {
```
High level doc comment plz

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2609,13 +2610,19 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 			return (true, NotifyOption::SkipPersist, Ok(()));
 		}
+		if !chan.is_outbound() {
+			if chan.maybe_trigger_autoclose_timer() {
```
I don't quite get why we start the autoclose timer here? Comment might help

valentinewallace commented on lightning/src/ln/channel.rs:
```diff
@@ -3163,6 +3195,22 @@ impl<Signer: Sign> Channel<Signer> {
+		}
+		if self.autoclose_timer == AUTOCLOSE_TIMEOUT {
+			// If the channel doesn't have pending HTLC outputs to claim on-chain
+			if self.pending_inbound_htlcs.len() + self.pending_outbound_htlcs.len() > 0 {
```
Should we be resetting the timer in the `else` case?

valentinewallace reviewed (commented):
Looks pretty reasonable! Will go more in-depth after the "require 25-50% feerate increase to force-close" updates

TheBlueMatt commented:
Moving to 0.1 as I'm not sure we should delay 0.0.101 or 0.0.102 for this.

TheBlueMatt commented:
Needs rebase.

ariard commented:
Rebased at f64c46c.

Few changes still last time :
* a new `should_autoclose` config value has been added to `ChannelConfig` (default: true), as the channel might have been opened with a trusted, always-online peer, thus making on-chain unilateral close irrelevant
* `AUTOCLOSE_TIMEOUT` has been scaled to 60 min (max lnd value), high enough to avoid closures with floppy peers but low enough to still have odds of block inclusion
* force-close are conditional on *included* htlcs being present, don't go on-chain for *dust* HTLCs

Especially happy to get opinions on a) `AUTOCLOSE_TIMEOUT` and b) sanitizing the inbound update fee being effectively upward in a given range (e.g 30%) to avoid malicious peers doing the strict minimum.

Still have to add test coverages.

TheBlueMatt commented on lightning/src/util/config.rs:
```diff
@@ -271,6 +276,7 @@ impl_writeable_tlv_based!(ChannelConfig, {
 	(4, announced_channel, required),
 	(6, commit_upfront_shutdown_pubkey, required),
 	(8, forwarding_fee_base_msat, required),
+	(10, should_autoclose, required),
```
This should be odd and have a `(default_value, X)` - no need to make old clients refuse to read new configs.

TheBlueMatt commented on lightning/src/util/config.rs:
```diff
@@ -246,6 +246,10 @@ pub struct ChannelConfig {
+	/// If this is set to false, we do not auto-close channel with pending time-sensitive outputs,
+	/// of which the feerate has been detected as stalling.
+	/// Default value: true.
+	pub should_autoclose: bool,
```
Maybe we should make this some kind of knob to configure how aggressively we do this instead? eg right now we require a feerate increase of 30%, maybe we should make that a knob and users who want to turn off autoclose can set it to `u64::max_value()`?

ariard replied:
What feerate we would like the commitment transaction to be committed with ? Ideally we would have visibility on the current mempool feerate groups composition to predict a worst-case evolution of those ones based on historical data. Assuming this worst-case, we should guarantee our feerate stay in-sync with the top feerate group for block inclusion during our `CLTV_CLAIM_BUFFER` (once we unilateraly break the chain we can't update it anymore). The current "30%" should represent our rate of fidelity of the worst-case, if a node operator would like to diverge from the worst-case, they should tweak it accordingly.                                                                                  

Even further, to have a more complete model, we should also tweak AUTOCLOSE_TIMEOUT to cover the velocity of the worst-case scenario ("Once we detected evolution towards unsafe mempool feerate groups,          
how fast we can react to correct the included feerate ?"). The issue being there, our reaction time is dependent on our counterparty interactivity latency, and there we could demand input from the node operator ("what's the liveliness reliability of this peer ?") to adjust in consequence AUTOCLOSE_TIMEOUT. The current AUTOCLOSE_TIMEOUT should be block-long enough to let our fee-estimator and our                    counterparty converge, but also short enough to match worst-case congestion velocity.

Ultimately the limitating safety factor is our counterparty latency. And that one can be quite extreme if our counterparty isn't robust against low-odds, high-impact risks (data center fire, DNS server outage, etc). Assuming package relay support and moving to anchor output as a fee-bumping mechanism remove this safety dependency on counterparty latency.

Or if you have a more rule-of-thumb thinking on wat to advice node operators to configure this potential knob ?

TheBlueMatt replied:
I don't think there is a clear answer, sadly. Indeed, no fee estimator (that I'm aware of) is really tuned for the lightning threat model. That said, the answer is likely some function over the trust a user has for their counterparty, which leads me to believe it should be a knob. If you're running a node that opens channels automatically with any counterparty, you probably want something pretty tight (30% seems reasonable?), if you're a node that only makes outbound channel opens with trusted peers, maybe 500% is more reasonable.

TheBlueMatt commented on lightning/src/ln/channel.rs:
```diff
@@ -2832,13 +2857,35 @@ impl<Signer: Sign> Channel<Signer> {
 	}
 
+	/// Trigger the autoclose timer if it's in the starting position
+	pub fn maybe_trigger_autoclose_timer(&mut self, current_feerate: u32, new_feerate: u32, background_feerate: u32) -> bool {
```
Why isn't this just a part of `send_update_fee`?

ariard replied:
There is at least a call in `update_channel_fee` to monitor stalling _inbound_ update_fee ?

TheBlueMatt replied:
Right, but this method is *indirectly* called in timer tick (via update_channel_fee) but `check_channel_autoclose` is called *directly* in timer tick. I still don't see a reason why they are separate, they're both called during timer ticks.

ariard replied:
I've verified the paths. I see, you're thinking to just pass a boolean `trigger_autoclose` to `send_update_fee`  down from `update_channel_fee`. IIRC, my motivation to have a separate method to start the auto-close timer was to make the mechanism re-usable on conditions beyond feerate variations such as peer disconnection, fee-bumping reserve exhaustion, etc.

TheBlueMatt commented on lightning/src/ln/channel.rs:
```diff
@@ -2832,13 +2857,35 @@ impl<Signer: Sign> Channel<Signer> {
+
+		// Verify the new feerate is at least superior by 30% of current feerate before to
+		// start a autoclose timer.
+		if new_feerate < current_feerate * 1300 / 1000 && new_feerate < background_feerate {
```
Lets check this on each timer tick? ie if feerate goes up some, but then comes back down, we should reset the timer to 0.

ariard replied:
Yes, done in `check_autoclose`.

ariard commented:
Updated at 285b0ec.

Note, removed the check against `background_feerate` in `maybe_trigger_autoclose`. Verifying the new feerate is above 30% of current channel feerate should be enough as fee upward detection check.

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3021,6 +3030,24 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
+		let mut notify_option = NotifyOption::SkipPersist;
+		let res = match chan.check_autoclose(new_feerate) {
+			Ok(res) => Ok(res),
+			Err(e) => {
```
Honestly this is a lot of error handling code, can we just inline this in timer_tick_occurred? It'd be a lot less total code.

TheBlueMatt commented:
One worry I have is what happens if we have an HTLC that times out in a day, but get no blocks for an hour and thus have a huge fee spike. One maybe simpler solution is to increase `CLTV_CLAIM_BUFFER` and `LATENCY_GRACE_PERIOD_BLOCKS` as used in `should_broadcast_holder_commitment_tx` by some ratio of how much feerate has increased compared to the latest holder commitment tx.

ariard commented:
Updated at eaabe00a with inbound-with-pending-payment-restriction-only and new config `autoclose_variation_rate`. 

> One maybe simpler solution is to increase CLTV_CLAIM_BUFFER and LATENCY_GRACE_PERIOD_BLOCKS as used in should_broadcast_holder_commitment_tx by some ratio of how much feerate has increased compared to the latest holder commitment tx.

I agree we could make `CLTV_CLAIM_BUFFER` and `LATENCY_GRACE_PERIOD_BLOCKS` dependent on fee spikes, extending them to provoke earlier closure of the channel and as such increasing the odds of transaction confirmation. Or we could also increase them in case of fee-bumping reserve exhaustion in the anchor output model. That said, I think there is another trade-off, if your fee-estimator gets compromised or unreliable, your channels are becoming far more frail, it might be the kind of mechanism where you don't want all the nodes behaving the same to avoid ecosystem-wide threshold effect.

TheBlueMatt commented:
> That said, I think there is another trade-off, if your fee-estimator gets compromised or unreliable, your channels are becoming far more frail, it might be the kind of mechanism where you don't want all the nodes behaving the same to avoid ecosystem-wide threshold effect.

Don't you have a similar trade-off with this PR as-is, though? In either case, if your fee estimator thinks the feerate needs to be higher, we'll force-close, even if the fee-estimator is garbage. At least moving this logic to the `ChannelMonitor` keeps our "should we force-close due to HTLC expiry" code in one place, and ensures a redundant `ChannelMonitor` can make the same decision.

I feel like doing it in that context also allows us a bit more specificity because we already have code around HTLC status there so its less total new code and we can bound stuff as we want.

ariard commented:
From online meeting : "accelerate the force-close of inbound HTLCs in case of slow-block issuance and feerate spikes, compress the block-timer in ChannelMonitor's `should_broadcast()`.

TheBlueMatt commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -1925,6 +1931,30 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
 						panic!("Attempted to replace shutdown script {} with {}", shutdown_script, scriptpubkey);
 					}
 				},
+				ChannelMonitorUpdateStep::FeesSpikes { feerate } => {
```
Why does this need a new Update type? Can't we just check the fee in ChannelMonitor when we call should_broadcast...?

ariard replied:
Because you still need a tick-based timer against which to build the staling block detection ? Could be in ChannelMonitor, but we need to add a new pub `timer_tick_occured` there, I think.

Otherwise what do you have in mind to detect block staling, or maybe you're thinking to compress `HTLC_CLAIM_BUFFER` on fees variations only ?

TheBlueMatt replied:
Wait, I'm confused, I thought we were roughly on the same page that we don't need to do this on a tick but can wait for the next block? There's a lot of complexity here that all goes away if we just wait for the next block, and I'm really unconvinced you're ever going to (a) hit the "fees are spiking and our channel fee is now too low" code *and* (b) get into the next block - if you got into the next block the "our channel fee is now too low" criteria almost certainly shouldn't have been hit.

ariard replied:
> Wait, I'm confused, I thought we were roughly on the same page that we don't need to do this on a tick but can wait for the next block?

Well, I'm confused too, I've noted "slow block issuance" as a criteria from meeting. And in that sense, I've implemented the timer as a poorman heuristic to detect slow block issuance. What I understand, is you suggest to rely on the fee spikes detection, as reported by our fee-estimator, to decide to compess `HTLC_CLAIM_BUFFER. If you assume to do it when we already hit `should_broadcast`, I think reduce the fault-tolerance worthiness of the feature, as actually you're waiting for the next block to get in, and potentially congestion to have worsened.

I'm good with that, though a) it should be knob-configurable to disable that mechanism if you trust your peers to be online and as thus reactive to your `update_fee` and b) a deprecation note to remove this mechanism once package relay is supported as it's kind of automatic-closure mechanism triggerable by a third-party owning your fee-estimator (sadly, a realistic assumption for our target phone users)

TheBlueMatt replied:
> If you assume to do it when we already hit should_broadcast, I think reduce the fault-tolerance worthiness of the feature, as actually you're waiting for the next block to get in, and potentially congestion to have worsened.

Right, that's true in theory, but I don't think it matters in practice, specifically my point that in order for this to be the case in practice, both of these need to be true:

> a) hit the "fees are spiking and our channel fee is now too low" code and (b) get into the next block - if you got into the next block the "our channel fee is now too low" criteria almost certainly shouldn't have been hit.

I don't think this code should/will ever trigger if the current channel fee qualifies for the next block. And if the current channel fee doesn't qualify for the next block, the whole debate seems moot to me?

ariard commented:
Updated at c6ca774

Simple force-close of inbound HTLCs at fee spikes detection in `should_broadcast_holder_commitment_txn`. Fee spikes rate is configurable on a knob.

TheBlueMatt commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -2662,7 +2679,7 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
 					let htlc_outbound = $holder_tx == htlc.offered;
 					if ( htlc_outbound && htlc.cltv_expiry + LATENCY_GRACE_PERIOD_BLOCKS <= height) ||
-					   (!htlc_outbound && htlc.cltv_expiry <= height + CLTV_CLAIM_BUFFER && self.payment_preimages.contains_key(&htlc.payment_hash)) {
+					   ((!htlc_outbound && (htlc.cltv_expiry <= height + CLTV_CLAIM_BUFFER || spikes_force_close)) && self.payment_preimages.contains_key(&htlc.payment_hash)) { //TODO: don't force-close if dust satoshis
```
Do we really want to force-close instantly if we only have one HTLC that expires in a day? I worry a bit about force-closes if a counterparty just uses a different fee estimator than us - fee estimators are really often 30+% off from each other.

ariard replied:
Added a new `get_fulfilled_htlc_sum()` as a new condition to determine if we force-close the chan or not. It doesn't account for dust HTLC of course.

ariard commented:
Rebased and completed at 5ae2cc7

Added test coverage.

ariard commented on lightning/src/chain/channelmonitor.rs:
```diff
@@ -2443,7 +2463,8 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
 		debug_assert!(self.best_block.height() >= conf_height);
 
-		let should_broadcast = self.should_broadcast_holder_commitment_txn(logger);
+		let fulfilled_htlc_high_exposure = 5_000_000 < self.get_fulfilled_htlc_sum();
```
Should we constify or make it a new config knob ?

I think we should give their say to node operators when it's question of risk evaluation. I pick up the same value than `max_dust_htlc_exposure`. Maybe we could pick one unified "funds at risk" value though at the same time it doesn't exactly cover the same risks...

Long-term, I think we should let node operators fine-tune their channels operations in function of their own risks analysis, instead of substituting to them (though likely too early).

ariard commented:
Closing for now : https://gnusha.org/bitcoin-rust/2022-03-27.log

While i think reaction dynamic on mempools congestion is the way to go long-term, `update_fee` as a fee-bumping mechanism is too limited, and it could be confusing or even harmful to users if we're closing early in unfavorable scenarios . To be revisited once we have anchor + package relay.
